pkg/auth/user/required_attribute_verifier.go

```diff
@@ -0,0 +1,39 @@
+package user
+
+import (
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/sliceutils"
+)
+
+// NewRequiredAttributesVerifier returns an AttributeVerifier that will verify all
+// attributes of a user and return an error if any of the required attributes are missing or
+// have a value that is different than expected.
+func NewRequiredAttributesVerifier(requiredAttributes []*storage.AuthProvider_RequiredAttribute) AttributeVerifier {
+	return &checkRequiredAttributesImpl{
+		required: requiredAttributes,
+	}
+}
+
+type checkRequiredAttributesImpl struct {
+	required []*storage.AuthProvider_RequiredAttribute
+}
+
+func (c checkRequiredAttributesImpl) Verify(attributes map[string][]string) error {
+	// User attributes do not _specifically_ have to be set, handle this explicitly.
+	if attributes == nil {
+		return errox.NoCredentials.CausedBy("none of the required attributes set")
+	}
+
+	for _, required := range c.required {
+		if observedValue, ok := attributes[required.GetAttributeKey()]; !ok || observedValue == nil {
+			// Explicitly return 401, as we do not want clients to be issued a token.
```

Comment:
Similar to the above, there are no clients and no tokens here. You _happen_ to know that `Verify()` will be called in such context(s) today but this is an incorrect assumption in general. I think all these errors shall be unclassified and the caller will wrap them in `errox.NoCredentials`. You won't be able to check for the error class in tests but all errors yielded here are of the same class so it's okay. (To be pedantic, the only caller of this function does not issue tokens either so you might need to adjust the comment there 😄).